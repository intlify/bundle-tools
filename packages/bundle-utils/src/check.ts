export type InstalledPackage = 'vue-i18n' | 'petite-vue-i18n'

// eslint-disable-next-line @typescript-eslint/ban-types
export function checkInstallPackage(
  pkg: string,
  debug: Function
): InstalledPackage {
  let installedVueI18n = false
  try {
    debug(`vue-i18n load path: ${require.resolve('vue-i18n')}`)
    installedVueI18n = true
  } catch (e) {
    debug(`cannot find 'vue-i18n'`, e)
  }

  let installedPetiteVueI18n = false
  try {
    debug(`petite-vue-i18n load path: ${require.resolve('petite-vue-i18n')}`)
    installedPetiteVueI18n = true
  } catch (e) {
    debug(`cannot find 'petite-vue-i18n'`, e)
  }

  if (installedVueI18n) {
    return 'vue-i18n'
  }
  if (installedPetiteVueI18n) {
    return 'petite-vue-i18n'
  }
  throw new Error(
    `${pkg} requires 'vue-i18n' or 'petite-vue-i18n' to be present in the dependency tree.`
  )
}

// eslint-disable-next-line @typescript-eslint/ban-types
export function checkVueI18nBridgeInstallPackage(debug: Function): boolean {
  let ret = false
  try {
    debug(`vue-i18n-bridge load path: ${require.resolve('vue-i18n-bridge')}`)
    ret = true
  } catch (e) {
    debug(`cannot find 'vue-i18n-bridge'`, e)
  }
  return ret
}
