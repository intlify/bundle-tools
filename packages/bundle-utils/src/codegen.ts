import module from 'node:module'
import {
  SourceMapGenerator,
  SourceMapConsumer,
  MappedPosition,
  MappingItem
} from 'source-map'
import { format, escapeHtml as sanitizeHtml, isBoolean } from '@intlify/shared'

import type { RawSourceMap } from 'source-map'
import type {
  CompileError,
  ResourceNode,
  CompileOptions
} from '@intlify/message-compiler'

// NOTE: use sourcemap (currently, `@intlify/message-compiler` support cjs pkg only)
const require = module.createRequire(import.meta.url)
const {
  baseCompile,
  detectHtmlTag,
  LocationStub
} = require('@intlify/message-compiler')

/**
 * Compilation dev environments
 *
 * @public
 */
export type DevEnv = 'development' | 'production'

export interface Position {
  line: number
  column: number
  offset?: number
}

export interface SourceLocationable {
  start?: number
  loc?: {
    start: Position
    end: Position
  } | null
}

/**
 * @internal
 */
export interface CodeGenOptions {
  type?: 'plain' | 'sfc' | 'bare'
  legacy?: boolean
  bridge?: boolean
  exportESM?: boolean
  source?: string
  sourceMap?: boolean
  filename?: string
  inSourceMap?: RawSourceMap
  isGlobal?: boolean
  locale?: string
  env?: DevEnv
  forceStringify?: boolean
  useClassComponent?: boolean
  allowDynamic?: boolean
  strictMessage?: boolean
  escapeHtml?: boolean
  onWarn?: (msg: string) => void
  onError?: (
    msg: string,
    extra?: {
      source: string
      path: string
      code?: CompileError['code']
      domain?: CompileError['domain']
      location?: CompileError['location']
    }
  ) => void
}

export interface CodeGenContext {
  source?: string
  code: string
  indentLevel: number
  filename: string
  line: number
  column: number
  offset: number
  env: DevEnv
  forceStringify: boolean
  map?: SourceMapGenerator
}

export interface CodeGenerator {
  context(): CodeGenContext
  push<Node extends SourceLocationable>(
    code: string,
    node?: Node,
    name?: string
  ): void
  indent(withNewLine?: boolean): void
  deindent(withNewLine?: boolean): void
  newline(): void
  pushline<Node extends SourceLocationable>(
    code: string,
    node?: Node,
    name?: string
  ): void
}

/**
 * @internal
 */
export interface CodeGenResult<ASTNode, CodeGenError extends Error = Error> {
  code: string
  ast: ASTNode
  errors?: CodeGenError[]
  map?: RawSourceMap
}

export function createCodeGenerator(
  options: CodeGenOptions = {
    filename: 'bundle.json',
    sourceMap: false,
    env: 'development',
    forceStringify: false
  }
): CodeGenerator {
  const { sourceMap, source, filename } = options
  const _context = Object.assign(
    {
      code: '',
      column: 1,
      line: 1,
      offset: 0,
      map: undefined,
      indentLevel: 0
    },
    options
  ) as CodeGenContext

  const context = (): CodeGenContext => _context

  function push<Node extends SourceLocationable>(
    code: string,
    node?: Node,
    name?: string
  ): void {
    _context.code += code
    if (_context.map && node) {
      if (node.loc && node.loc !== LocationStub) {
        addMapping(node.loc.start, name)
      }
      advancePositionWithSource(_context as Position, code)
    }
  }

  function _newline(n: number): void {
    push('\n' + `  `.repeat(n))
  }

  function indent(withNewLine = true): void {
    const level = ++_context.indentLevel
    withNewLine && _newline(level)
  }

  function deindent(withNewLine = true): void {
    const level = --_context.indentLevel
    withNewLine && _newline(level)
  }

  function newline(): void {
    _newline(_context.indentLevel)
  }

  function pushline<Node extends SourceLocationable>(
    code: string,
    node?: Node,
    name?: string
  ): void {
    push(code, node, name)
    newline()
  }

  function addMapping(loc: Position, name?: string) {
    _context.map!.addMapping({
      name,
      source: _context.filename,
      original: {
        line: loc.line,
        column: loc.column - 1
      },
      generated: {
        line: _context.line,
        column: _context.column - 1
      }
    })
  }

  if (sourceMap && source) {
    _context.map = new SourceMapGenerator()
    _context.map.setSourceContent(filename!, source)
  }

  return {
    context,
    push,
    indent,
    deindent,
    newline,
    pushline
  }
}

function advancePositionWithSource(
  pos: Position,
  source: string,
  numberOfCharacters: number = source.length
): Position {
  if (pos.offset == null) {
    return pos
  }

  let linesCount = 0
  let lastNewLinePos = -1
  for (let i = 0; i < numberOfCharacters; i++) {
    if (source.charCodeAt(i) === 10 /* newline char code */) {
      linesCount++
      lastNewLinePos = i
    }
  }

  pos.offset += numberOfCharacters
  pos.line += linesCount
  pos.column =
    lastNewLinePos === -1
      ? pos.column + numberOfCharacters
      : numberOfCharacters - lastNewLinePos

  return pos
}

const DETECT_MESSAGE = `Detected HTML in '{msg}' message.`
const ON_ERROR_NOOP = () => {} // eslint-disable-line @typescript-eslint/no-empty-function

function parsePath(path?: string[]): string {
  return path ? path.join('.') : ''
}

export function generateMessageFunction(
  msg: string,
  options: CodeGenOptions,
  path?: string[]
): CodeGenResult<ResourceNode> {
  const env = options.env != null ? options.env : 'development'
  const strictMessage = isBoolean(options.strictMessage)
    ? options.strictMessage
    : true
  const escapeHtml = !!options.escapeHtml
  const onError = options.onError || ON_ERROR_NOOP
  const errors = [] as CompileError[]

  let detecteHtmlInMsg = false
  if (detectHtmlTag(msg)) {
    detecteHtmlInMsg = true
    if (strictMessage) {
      const errMsg = format(DETECT_MESSAGE, { msg })
      onError(format(errMsg), {
        source: msg,
        path: parsePath(path)
      })
    }
  }

  const _msg = detecteHtmlInMsg && escapeHtml ? sanitizeHtml(msg) : msg

  const newOptions = Object.assign({ mode: 'arrow' }, options) as CompileOptions
  newOptions.onError = (err: CompileError): void => {
    if (onError) {
      const extra: Parameters<Required<CodeGenOptions>['onError']>[1] = {
        source: msg,
        path: parsePath(path),
        code: err.code,
        domain: err.domain,
        location: err.location
      }
      onError(err.message, extra)
      errors.push(err)
    }
  }
  const { code, ast, map } = baseCompile(_msg, newOptions)
  const occured = errors.length > 0
  const genCode = !occured
    ? env === 'development'
      ? `(()=>{const fn=${code};fn.source=${JSON.stringify(msg)};return fn;})()`
      : `${code}`
    : _msg
  return { code: genCode, ast, map, errors }
}

export function mapLinesColumns(
  resMap: RawSourceMap,
  codeMaps: Map<string, RawSourceMap>,
  inSourceMap?: RawSourceMap
): RawSourceMap | null {
  if (!resMap) {
    return null
  }

  const resMapConsumer = new SourceMapConsumer(resMap)
  const inMapConsumer = inSourceMap ? new SourceMapConsumer(inSourceMap) : null
  const mergedMapGenerator = new SourceMapGenerator()

  let inMapFirstItem: MappingItem | null = null
  if (inMapConsumer) {
    inMapConsumer.eachMapping(m => {
      if (inMapFirstItem) {
        return
      }
      inMapFirstItem = m
    })
  }

  resMapConsumer.eachMapping(res => {
    if (res.originalLine == null) {
      return
    }

    const map = codeMaps.get(res.name)
    if (!map) {
      return
    }

    let inMapOrigin: MappedPosition | null = null
    if (inMapConsumer) {
      inMapOrigin = inMapConsumer.originalPositionFor({
        line: res.originalLine,
        column: res.originalColumn - 1
      })
      if (inMapOrigin.source == null) {
        inMapOrigin = null
        return
      }
    }

    const mapConsumer = new SourceMapConsumer(map)
    mapConsumer.eachMapping(m => {
      mergedMapGenerator.addMapping({
        original: {
          line: inMapFirstItem
            ? inMapFirstItem.originalLine + res.originalLine - 2
            : res.originalLine,
          column: inMapFirstItem
            ? inMapFirstItem.originalColumn + res.originalColumn
            : res.originalColumn
        },
        generated: {
          line: inMapFirstItem
            ? inMapFirstItem.generatedLine + res.originalLine - 2
            : res.originalLine,
          // map column with message format compilation code map
          column: inMapFirstItem
            ? inMapFirstItem.generatedColumn +
              res.originalColumn +
              m.generatedColumn
            : res.originalColumn + m.generatedColumn
        },
        source: inMapOrigin ? inMapOrigin.source : res.source,
        name: m.name // message format compilation code
      })
    })
  })

  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  const generator = mergedMapGenerator as any
  // const targetConsumer = inMapConsumer || resMapConsumer
  const targetConsumer = inMapConsumer || resMapConsumer
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  ;(targetConsumer as any).sources.forEach((sourceFile: string) => {
    generator._sources.add(sourceFile)
    const sourceContent = targetConsumer.sourceContentFor(sourceFile)
    if (sourceContent != null) {
      mergedMapGenerator.setSourceContent(sourceFile, sourceContent)
    }
  })

  generator._sourceRoot = inSourceMap
    ? inSourceMap.sourceRoot
    : resMap.sourceRoot
  generator._file = inSourceMap ? inSourceMap.file : resMap.file

  return generator.toJSON()
}
