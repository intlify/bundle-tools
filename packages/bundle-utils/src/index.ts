export { generate as generateJSON } from './json'
export { generate as generateYAML } from './yaml'
export { generate as generateJavaScript } from './js'
export {
  getVueI18nVersion,
  checkInstallPackage,
  checkVueI18nBridgeInstallPackage
} from './deps'
export type { CodeGenOptions, CodeGenResult, DevEnv } from './codegen'
export type { InstalledPackage } from './deps'
