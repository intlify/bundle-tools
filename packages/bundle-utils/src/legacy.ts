import { friendlyJSONstringify } from '@intlify/shared'

import type { CodeGenOptions } from './codegen'

export function generateLegacyCode(
  { isGlobal }: CodeGenOptions,
  generator: () => string
) {
  // prettier-ignore
  const componentNamespace = `Component.options`
  const variableName = !isGlobal ? '__i18n' : '__i18nGlobal'
  const exportSyntax = 'export default'
  const code = `${exportSyntax} function (Component) {
  ${componentNamespace}.${variableName} = ${componentNamespace}.${variableName} || []
  ${componentNamespace}.${variableName}.push(${friendlyJSONstringify(
    generator()
  )})
  delete ${componentNamespace}._Ctor
}`
  return code
}
