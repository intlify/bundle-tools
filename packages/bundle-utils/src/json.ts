/**
 * Code generator for i18n json/json5 resource
 */

import {
  parseJSON,
  traverseNodes,
  getStaticJSONValue
} from 'jsonc-eslint-parser'
import { isString, friendlyJSONstringify } from '@intlify/shared'
import {
  createCodeGenerator,
  generateMessageFunction,
  mapLinesColumns
} from './codegen'
import { generateLegacyCode } from './legacy'
import { RawSourceMap } from 'source-map'
import MagicString from 'magic-string'

import type { JSONProgram, JSONNode } from 'jsonc-eslint-parser/lib/parser/ast'
import type { CodeGenOptions, CodeGenerator, CodeGenResult } from './codegen'

/**
 * @internal
 */
export function generate(
  targetSource: string | Buffer,
  {
    type = 'plain',
    legacy = false,
    bridge = false,
    exportESM = false,
    filename = 'vue-i18n-loader.json',
    inSourceMap = undefined,
    locale = '',
    isGlobal = false,
    sourceMap = false,
    env = 'development',
    forceStringify = false,
    onError = undefined,
    strictMessage = true,
    escapeHtml = false,
    useClassComponent = false
  }: CodeGenOptions,
  injector?: () => string
): CodeGenResult<JSONProgram> {
  const target = Buffer.isBuffer(targetSource)
    ? targetSource.toString()
    : targetSource
  // const value = JSON.stringify(JSON.parse(target))
  //   .replace(/\u2028/g, '\\u2028') // line separator
  //   .replace(/\u2029/g, '\\u2029') // paragraph separator
  const value = target

  const options = {
    type,
    bridge,
    exportESM,
    source: value,
    sourceMap,
    locale,
    isGlobal,
    inSourceMap,
    env,
    filename,
    forceStringify,
    onError,
    strictMessage,
    escapeHtml,
    useClassComponent
  } as CodeGenOptions
  const generator = createCodeGenerator(options)

  const ast = parseJSON(value, { filePath: filename })

  // for vue 2.x
  if (legacy && type === 'sfc') {
    const gen = () => friendlyJSONstringify(getStaticJSONValue(ast))
    const code = generateLegacyCode(options, gen)
    const s = new MagicString(code)
    return {
      ast,
      code: s.toString(),
      map: s.generateMap({
        file: filename,
        source: value,
        includeContent: true
      }) as unknown as RawSourceMap
    }
  }

  const codeMaps = generateNode(generator, ast, options, injector)

  const { code, map } = generator.context()
  // if (map) {
  //   const s = new SourceMapConsumer((map as any).toJSON())
  //   s.eachMapping(m => {
  //     console.log('sourcemap json', m)
  //   })
  // }
  // prettier-ignore
  const newMap = map
    ? mapLinesColumns((map as any).toJSON(), codeMaps, inSourceMap) || null // eslint-disable-line @typescript-eslint/no-explicit-any
    : null
  return {
    ast,
    code,
    map: newMap != null ? newMap : undefined
  }
}

function generateNode(
  generator: CodeGenerator,
  node: JSONProgram,
  options: CodeGenOptions,
  injector?: () => string
): Map<string, RawSourceMap> {
  const propsCountStack = [] as number[]
  const pathStack = [] as string[]
  const itemsCountStack = [] as number[]
  const { forceStringify } = generator.context()
  const codeMaps = new Map<string, RawSourceMap>()
  const {
    type,
    bridge,
    exportESM,
    sourceMap,
    isGlobal,
    locale,
    useClassComponent
  } = options

  const componentNamespace = '_Component'

  traverseNodes(node, {
    enterNode(node: JSONNode, parent: JSONNode) {
      switch (node.type) {
        case 'Program':
          if (type === 'plain') {
            generator.push(`const resource = `)
          } else if (type === 'sfc') {
            // for 'sfc'
            const variableName =
              type === 'sfc' ? (!isGlobal ? '__i18n' : '__i18nGlobal') : ''
            const localeName =
              type === 'sfc' ? (locale != null ? locale : `""`) : ''
            const exportSyntax = bridge
              ? exportESM
                ? `export default`
                : `module.exports =`
              : `export default`
            generator.push(`${exportSyntax} function (Component) {`)
            generator.indent()
            // prettier-ignore
            const componentVariable = bridge
              ? `Component.options || Component`
              : useClassComponent
                ? `Component.__o || Component.__vfdConstructor || Component`
                : `Component`
            // prettier-ignore
            generator.pushline(`const ${componentNamespace} = ${componentVariable}`)
            generator.pushline(
              `${componentNamespace}.${variableName} = ${componentNamespace}.${variableName} || []`
            )
            generator.push(`${componentNamespace}.${variableName}.push({`)
            generator.indent()
            generator.pushline(`"locale": ${JSON.stringify(localeName)},`)
            generator.push(`"resource": `)
          }
          break
        case 'JSONObjectExpression':
          generator.push(`{`)
          generator.indent()
          propsCountStack.push(node.properties.length)
          if (parent.type === 'JSONArrayExpression') {
            const lastIndex = itemsCountStack.length - 1
            const currentCount =
              parent.elements.length - itemsCountStack[lastIndex]
            pathStack.push(currentCount.toString())
            itemsCountStack[lastIndex] = --itemsCountStack[lastIndex]
          }
          break
        case 'JSONProperty':
          if (
            node.value.type === 'JSONLiteral' &&
            (node.key.type === 'JSONLiteral' ||
              node.key.type === 'JSONIdentifier')
          ) {
            const name =
              node.key.type === 'JSONLiteral' ? node.key.value : node.key.name
            const value = node.value.value
            if (isString(value)) {
              generator.push(`${JSON.stringify(name)}: `)
              pathStack.push(name.toString())
              const { code, map } = generateMessageFunction(
                value,
                options,
                pathStack
              )
              sourceMap && map != null && codeMaps.set(value, map)
              generator.push(`${code}`, node.value, value)
            } else {
              if (forceStringify) {
                const strValue = JSON.stringify(value)
                generator.push(`${JSON.stringify(name)}: `)
                pathStack.push(name.toString())
                const { code, map } = generateMessageFunction(
                  strValue,
                  options,
                  pathStack
                )
                sourceMap && map != null && codeMaps.set(strValue, map)
                generator.push(`${code}`, node.value, strValue)
              } else {
                generator.push(
                  `${JSON.stringify(name)}: ${JSON.stringify(value)}`
                )
                pathStack.push(name.toString())
              }
            }
          } else if (
            (node.value.type === 'JSONObjectExpression' ||
              node.value.type === 'JSONArrayExpression') &&
            (node.key.type === 'JSONLiteral' ||
              node.key.type === 'JSONIdentifier')
          ) {
            const name =
              node.key.type === 'JSONLiteral' ? node.key.value : node.key.name
            generator.push(`${JSON.stringify(name)}: `)
            pathStack.push(name.toString())
          }
          const lastIndex = propsCountStack.length - 1
          propsCountStack[lastIndex] = --propsCountStack[lastIndex]
          break
        case 'JSONArrayExpression':
          generator.push(`[`)
          generator.indent()
          if (parent.type === 'JSONArrayExpression') {
            const lastIndex = itemsCountStack.length - 1
            const currentCount =
              parent.elements.length - itemsCountStack[lastIndex]
            pathStack.push(currentCount.toString())
            itemsCountStack[lastIndex] = --itemsCountStack[lastIndex]
          }
          itemsCountStack.push(node.elements.length)
          break
        case 'JSONLiteral':
          if (parent.type === 'JSONArrayExpression') {
            const lastIndex = itemsCountStack.length - 1
            const currentCount =
              parent.elements.length - itemsCountStack[lastIndex]
            pathStack.push(currentCount.toString())
            if (node.type === 'JSONLiteral') {
              const value = node.value
              if (isString(value)) {
                const { code, map } = generateMessageFunction(
                  value,
                  options,
                  pathStack
                )
                sourceMap && map != null && codeMaps.set(value, map)
                generator.push(`${code}`, node, value)
              } else {
                if (forceStringify) {
                  const strValue = JSON.stringify(value)
                  const { code, map } = generateMessageFunction(
                    strValue,
                    options,
                    pathStack
                  )
                  sourceMap && map != null && codeMaps.set(strValue, map)
                  generator.push(`${code}`, node, strValue)
                } else {
                  generator.push(`${JSON.stringify(value)}`)
                }
              }
            }
            itemsCountStack[lastIndex] = --itemsCountStack[lastIndex]
          }
          break
        default:
          break
      }
    },
    leaveNode(node: JSONNode, parent: JSONNode) {
      switch (node.type) {
        case 'Program':
          if (type === 'sfc') {
            generator.deindent()
            generator.push(`})`)
            if (bridge && injector) {
              generator.newline()
              generator.pushline(
                `${componentNamespace}.__i18nBridge = ${componentNamespace}.__i18nBridge || []`
              )
              generator.pushline(
                `${componentNamespace}.__i18nBridge.push('${injector()}')`
              )
              generator.pushline(`delete ${componentNamespace}._Ctor`)
            }
            generator.deindent()
            generator.pushline(`}`)
          } else if (type === 'plain') {
            generator.push(`\n`)
            generator.push('export default resource')
          }
          break
        case 'JSONObjectExpression':
          if (propsCountStack[propsCountStack.length - 1] === 0) {
            pathStack.pop()
            propsCountStack.pop()
          }
          generator.deindent()
          generator.push(`}`)
          if (parent.type === 'JSONArrayExpression') {
            if (itemsCountStack[itemsCountStack.length - 1] !== 0) {
              pathStack.pop()
              generator.pushline(`,`)
            }
          }
          break
        case 'JSONProperty':
          if (propsCountStack[propsCountStack.length - 1] !== 0) {
            pathStack.pop()
            generator.pushline(`,`)
          }
          break
        case 'JSONArrayExpression':
          if (itemsCountStack[itemsCountStack.length - 1] === 0) {
            pathStack.pop()
            itemsCountStack.pop()
          }
          generator.deindent()
          generator.push(`]`)
          if (parent.type === 'JSONArrayExpression') {
            if (itemsCountStack[itemsCountStack.length - 1] !== 0) {
              pathStack.pop()
              generator.pushline(`,`)
            }
          }
          break
        case 'JSONLiteral':
          if (parent.type === 'JSONArrayExpression') {
            if (itemsCountStack[itemsCountStack.length - 1] !== 0) {
              pathStack.pop()
              generator.pushline(`,`)
            } else {
              generator.pushline(`,`)
            }
          }
          break
        default:
          break
      }
    }
  })

  return codeMaps
}
