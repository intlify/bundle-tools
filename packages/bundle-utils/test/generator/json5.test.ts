import { readFile } from '../utils'
import { generate } from '../../src/json'

test('json5', async () => {
  const { source } = await readFile('./fixtures/codegen/complex.json5')
  const { code, map } = generate(source, {
    sourceMap: true,
    env: 'development'
  })

  expect(code).toMatchSnapshot('code')
  expect(map).toMatchSnapshot('map')
})

test('AST code generation', async () => {
  const { source } = await readFile('./fixtures/codegen/complex.json5')
  const { code } = generate(source, {
    jit: true,
    env: 'production'
  })

  expect(code).toMatchSnapshot('code')
})
