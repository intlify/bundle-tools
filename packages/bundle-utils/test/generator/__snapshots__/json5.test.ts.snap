// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`AST code generation > code 1`] = `
"const resource = {
  \\"hi\\": {\\"type\\":0,\\"body\\":{\\"type\\":2,\\"items\\":[{\\"type\\":3,\\"value\\":\\"hi there!\\"}],\\"static\\":\\"hi there!\\"}},
  \\"nested\\": {
    \\"hello\\": {\\"type\\":0,\\"body\\":{\\"type\\":2,\\"items\\":[{\\"type\\":3,\\"value\\":\\"hello world!\\"}],\\"static\\":\\"hello world!\\"}},
    \\"more\\": {
      \\"plural\\": {\\"type\\":0,\\"body\\":{\\"type\\":1,\\"cases\\":[{\\"type\\":2,\\"items\\":[{\\"type\\":6,\\"modifier\\":{\\"type\\":8,\\"value\\":\\"caml\\"},\\"key\\":{\\"type\\":9,\\"value\\":\\"no apples\\"}}]},{\\"type\\":2,\\"items\\":[{\\"type\\":5,\\"index\\":0},{\\"type\\":3,\\"value\\":\\" apple\\"}]},{\\"type\\":2,\\"items\\":[{\\"type\\":4,\\"key\\":\\"n\\"},{\\"type\\":3,\\"value\\":\\" apples\\"}]}]}}
    },
    \\"list\\": {\\"type\\":0,\\"body\\":{\\"type\\":2,\\"items\\":[{\\"type\\":3,\\"value\\":\\"hi, \\"},{\\"type\\":5,\\"index\\":0},{\\"type\\":3,\\"value\\":\\" !\\"}]}}
  },
  \\"„Åì„Çì„Å´„Å°„ÅØ\\": {\\"type\\":0,\\"body\\":{\\"type\\":2,\\"items\\":[{\\"type\\":3,\\"value\\":\\"„Åì„Çì„Å´„Å°„ÅØÔºÅ\\"}],\\"static\\":\\"„Åì„Çì„Å´„Å°„ÅØÔºÅ\\"}},
  \\"single-quote\\": {\\"type\\":0,\\"body\\":{\\"type\\":2,\\"items\\":[{\\"type\\":3,\\"value\\":\\"I don\\\\u0027t know!\\"}],\\"static\\":\\"I don\\\\u0027t know!\\"}},
  \\"emoji\\": {\\"type\\":0,\\"body\\":{\\"type\\":2,\\"items\\":[{\\"type\\":3,\\"value\\":\\"üò∫\\"}],\\"static\\":\\"üò∫\\"}},
  \\"unicode\\": {\\"type\\":0,\\"body\\":{\\"type\\":2,\\"items\\":[{\\"type\\":3,\\"value\\":\\"A\\"}],\\"static\\":\\"A\\"}},
  \\"unicode-escape\\": {\\"type\\":0,\\"body\\":{\\"type\\":2,\\"items\\":[{\\"type\\":3,\\"value\\":\\"\\\\\\\\u0041\\"}],\\"static\\":\\"\\\\\\\\u0041\\"}},
  \\"backslash-single-quote\\": {\\"type\\":0,\\"body\\":{\\"type\\":2,\\"items\\":[{\\"type\\":3,\\"value\\":\\"\\\\\\\\\\\\u0027\\"}],\\"static\\":\\"\\\\\\\\\\\\u0027\\"}},
  \\"backslash-backslash\\": {\\"type\\":0,\\"body\\":{\\"type\\":2,\\"items\\":[{\\"type\\":3,\\"value\\":\\"\\\\\\\\\\\\\\\\\\"}],\\"static\\":\\"\\\\\\\\\\\\\\\\\\"}},
  \\"errors\\": [
    {\\"type\\":0,\\"body\\":{\\"type\\":2,\\"items\\":[{\\"type\\":3,\\"value\\":\\"ERROR1001\\"}],\\"static\\":\\"ERROR1001\\"}},
    {\\"type\\":0,\\"body\\":{\\"type\\":2,\\"items\\":[{\\"type\\":3,\\"value\\":\\"ERROR1002\\"}],\\"static\\":\\"ERROR1002\\"}},
    
  ],
  \\"complex\\": {
    \\"warnings\\": [
      {\\"type\\":0,\\"body\\":{\\"type\\":2,\\"items\\":[{\\"type\\":3,\\"value\\":\\"NOTE: This is warning\\"}],\\"static\\":\\"NOTE: This is warning\\"}},
      {
        \\"named-waring\\": {\\"type\\":0,\\"body\\":{\\"type\\":2,\\"items\\":[{\\"type\\":3,\\"value\\":\\"this is \\"},{\\"type\\":4,\\"key\\":\\"type\\"},{\\"type\\":3,\\"value\\":\\" warining\\"}]}}
      }
    ]
  }
}
export default resource"
`;

exports[`json5 > code 1`] = `
"const resource = {
  \\"hi\\": (()=>{const fn=(ctx) => {const { normalize: _normalize } = ctx;return _normalize([\\"hi there!\\"])};fn.source=\\"hi there!\\";return fn;})(),
  \\"nested\\": {
    \\"hello\\": (()=>{const fn=(ctx) => {const { normalize: _normalize } = ctx;return _normalize([\\"hello world!\\"])};fn.source=\\"hello world!\\";return fn;})(),
    \\"more\\": {
      \\"plural\\": (()=>{const fn=(ctx) => {const { normalize: _normalize, linked: _linked, type: _type, interpolate: _interpolate, list: _list, named: _named, plural: _plural } = ctx;return _plural([_normalize([_linked(\\"no apples\\", \\"caml\\", _type)]), _normalize([_interpolate(_list(0)), \\" apple\\"]), _normalize([_interpolate(_named(\\"n\\")), \\" apples\\"])])};fn.source=\\"@.caml:{'no apples'} | {0} apple | {n} apples\\";return fn;})()
    },
    \\"list\\": (()=>{const fn=(ctx) => {const { normalize: _normalize, interpolate: _interpolate, list: _list } = ctx;return _normalize([\\"hi, \\", _interpolate(_list(0)), \\" !\\"])};fn.source=\\"hi, {0} !\\";return fn;})()
  },
  \\"„Åì„Çì„Å´„Å°„ÅØ\\": (()=>{const fn=(ctx) => {const { normalize: _normalize } = ctx;return _normalize([\\"„Åì„Çì„Å´„Å°„ÅØÔºÅ\\"])};fn.source=\\"„Åì„Çì„Å´„Å°„ÅØÔºÅ\\";return fn;})(),
  \\"single-quote\\": (()=>{const fn=(ctx) => {const { normalize: _normalize } = ctx;return _normalize([\\"I don't know!\\"])};fn.source=\\"I don't know!\\";return fn;})(),
  \\"emoji\\": (()=>{const fn=(ctx) => {const { normalize: _normalize } = ctx;return _normalize([\\"üò∫\\"])};fn.source=\\"üò∫\\";return fn;})(),
  \\"unicode\\": (()=>{const fn=(ctx) => {const { normalize: _normalize } = ctx;return _normalize([\\"A\\"])};fn.source=\\"A\\";return fn;})(),
  \\"unicode-escape\\": (()=>{const fn=(ctx) => {const { normalize: _normalize } = ctx;return _normalize([\\"\\\\\\\\u0041\\"])};fn.source=\\"\\\\\\\\u0041\\";return fn;})(),
  \\"backslash-single-quote\\": (()=>{const fn=(ctx) => {const { normalize: _normalize } = ctx;return _normalize([\\"\\\\\\\\'\\"])};fn.source=\\"\\\\\\\\'\\";return fn;})(),
  \\"backslash-backslash\\": (()=>{const fn=(ctx) => {const { normalize: _normalize } = ctx;return _normalize([\\"\\\\\\\\\\\\\\\\\\"])};fn.source=\\"\\\\\\\\\\\\\\\\\\";return fn;})(),
  \\"errors\\": [
    (()=>{const fn=(ctx) => {const { normalize: _normalize } = ctx;return _normalize([\\"ERROR1001\\"])};fn.source=\\"ERROR1001\\";return fn;})(),
    (()=>{const fn=(ctx) => {const { normalize: _normalize } = ctx;return _normalize([\\"ERROR1002\\"])};fn.source=\\"ERROR1002\\";return fn;})(),
    
  ],
  \\"complex\\": {
    \\"warnings\\": [
      (()=>{const fn=(ctx) => {const { normalize: _normalize } = ctx;return _normalize([\\"NOTE: This is warning\\"])};fn.source=\\"NOTE: This is warning\\";return fn;})(),
      {
        \\"named-waring\\": (()=>{const fn=(ctx) => {const { normalize: _normalize, interpolate: _interpolate, named: _named } = ctx;return _normalize([\\"this is \\", _interpolate(_named(\\"type\\")), \\" warining\\"])};fn.source=\\"this is {type} warining\\";return fn;})()
      }
    ]
  }
}
export default resource"
`;

exports[`json5 > map 1`] = `
{
  "mappings": ";wEACKA;;6EAEKC;;uMAEGC,aAAAC,8BAAAC,wBAAAC,wBAAAC,2BAAAC;;oHAEJC,QAAAJ,wBAAAK;;;6EAGCC;oFACOC;2EACTC;6EACEC;sFACSC;8FACQC;2FACHC;;sEAErBC,aAAaC;;;;wEAIXC;;oIAEkBC,YAAAC,8BAAAC",
  "names": [
    "hi there!",
    "hello world!",
    "no apples",
    "caml",
    "0",
    " apple",
    "n",
    " apples",
    "hi, ",
    " !",
    "„Åì„Çì„Å´„Å°„ÅØÔºÅ",
    "I don't know!",
    "üò∫",
    "A",
    "\\\\u0041",
    "\\\\'",
    "\\\\\\\\",
    "ERROR1001",
    "ERROR1002",
    "NOTE: This is warning",
    "this is ",
    "type",
    " warining",
  ],
  "sources": [
    "vue-i18n-loader.json",
  ],
  "sourcesContent": [
    "{
  hi: \\"hi there!\\",
  nested: {
    hello: \\"hello world!\\",
    more: {
      plural: \\"@.caml:{'no apples'} | {0} apple | {n} apples\\"
    },
    list: \\"hi, {0} !\\"
  },
  // comment
  \\"„Åì„Çì„Å´„Å°„ÅØ\\": \\"„Åì„Çì„Å´„Å°„ÅØÔºÅ\\",
  \\"single-quote\\": \\"I don't know!\\",
  emoji: \\"üò∫\\",
  unicode: \\"\\\\u0041\\",
  \\"unicode-escape\\": \\"\\\\\\\\u0041\\",
  \\"backslash-single-quote\\": \\"\\\\\\\\'\\",
  \\"backslash-backslash\\": \\"\\\\\\\\\\\\\\\\\\",
  errors: [
    'ERROR1001', 'ERROR1002'
  ],
  complex: {
    warnings: [
      'NOTE: This is warning',
      {
        \\"named-waring\\": 'this is {type} warining'
      }
    ]
  }
}
",
  ],
  "version": 3,
}
`;
